flink-dist/src/main/flink-bin/bin/config.sh

```diff
@@ -645,3 +645,20 @@ extractExecutionParams() {
 
     echo ${execution_config} | sed "s/$EXECUTION_PREFIX//"
 }
+
+findLog4jConfiguration() {
+    local PATH=$1
+    local LOG4J_CONFIG_FILES=($PATH/log4j-console.*)
+    if [[ ${#LOG4J_CONFIG_FILES[@]} -gt 1 ]]; then
+        echo "[ERROR] Found more than one log4j configuration file: [${LOG4J_CONFIG_FILES[@]}]" 1>&2
+        exit 1
+    fi
+
+    LOG4J_CONFIG=${LOG4J_CONFIG_FILES[0]}
+    if [[ "${LOG4J_CONFIG: -1}" = "*" ]]; then
+        echo "[ERROR] Could not find a valid configuration" 1>&2
+        exit 1
+    fi
```

Comment:
I don't understand this block. If we wanted to check whether there were any `log4j-console.*` files, couldn't we check for an empty array in the previous if-block?

What does `if [[ "${LOG4J_CONFIG: -1}" = "*" ]]; then` test?

Reply:
What if the user uses logback, will this fail then?

Reply:
(Context: We use logback and do not have a log4j file).

Reply:
> I don't understand this block. If we wanted to check whether there were any `log4j-console.*` files, couldn't we check for an empty array in the previous if-block?
> 
> What does `if [[ "${LOG4J_CONFIG: -1}" = "*" ]]; then` test?

The problem, I experienced, is that the query `($PATH/log4j-console.*)` returns itself if no matches were found. This means the array will never be empty and therefore the additional check.

> What if the user uses logback, will this fail then?

It should not be affected right now since it is set through a different java configuration setting `-Dlogback.configurationFile`.

Reply:
> The problem, I experienced, is that the query `($PATH/log4j-console.*)` returns itself if no matches were found. This means the array will never be empty and therefore the additional check.

Thanks for clarifying, I think it wouldn't hurt to add a comment to explain that bash behavior.

> It should not be affected right now since it is set through a different java configuration setting `-Dlogback.configurationFile`.

I believe it would fail because the changes _require_ a log4j file to be present.

Reply:
Hmm, I am not familiar with the logback setup but I think these changes do not require more presence of a log4j file than before.

The `findLog4jConfiguration()` call is only used in a subshell which will `exit 1` and does not affect the main shell.  Although I agree the log message `[ERROR] Could not find a valid configuration` is not ideal because, in the case of the logback usage, it is probably intended not to have a log4j configuration. 
It might be better to display a warning rather than an error that users are aware no log4j configuration could be found and it falls back to something else.

Reply:
I didn't see that the function call was inside backticks which lets it run in a subshell. Just making sure the log4j configuration remains optional.

Should we return a default value in case of an error? Otherwise these will be the resulting parameters:

```
-Dlog4j.configuration=file: -Dlog4j.configurationFile=file:
```

Not sure if that could cause an error.